osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs

```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
+        /// If it is before the first frame of the replay or the after the last frame of the replay, <see cref="CurrentFrame"/> and <see cref="NextFrame"/> agree.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The replay is empty.</exception>
         public TFrame NextFrame
         {
             get
             {
                 if (!HasFrames)
-                    return null;
-
-                if (!currentFrameIndex.HasValue)
-                    return currentDirection > 0 ? (TFrame)Frames[0] : null;
-
-                int nextFrame = clampedNextFrameIndex;
-
-                if (nextFrame == currentFrameIndex.Value)
-                    return null;
+                    throw new InvalidOperationException($"Attempted to get {nameof(NextFrame)} of an empty replay");
 
-                return (TFrame)Frames[clampedNextFrameIndex];
+                return (TFrame)Frames[Math.Min(currentFrameIndex + 1, Frames.Count - 1)];
             }
         }
 
-        private int? currentFrameIndex;
+        /// <summary>
+        /// When set, we will ensure frames executed by nested drawables are frame-accurate to replay data.
+        /// Disabling this can make replay playback smoother (useful for autoplay, currently).
+        /// </summary>
+        public bool FrameAccuratePlayback;
 
-        private int clampedNextFrameIndex =>
-            currentFrameIndex.HasValue ? Math.Clamp(currentFrameIndex.Value + currentDirection, 0, Frames.Count - 1) : 0;
+        // This input handler should be enabled only if there is at least one replay frame.
+        public override bool IsActive => HasFrames;
 
-        protected FramedReplayInputHandler(Replay replay)
-        {
-            this.replay = replay;
-        }
-
-        private const double sixty_frame_time = 1000.0 / 60;
+        // Can make it non-null but that is a breaking change.
+        protected double? CurrentTime { get; private set; }
 
         protected virtual double AllowedImportantTimeSpan => sixty_frame_time * 1.2;
 
-        protected double? CurrentTime { get; private set; }
+        protected List<ReplayFrame> Frames => replay.Frames;
 
-        private int currentDirection = 1;
+        private readonly Replay replay;
 
-        /// <summary>
-        /// When set, we will ensure frames executed by nested drawables are frame-accurate to replay data.
-        /// Disabling this can make replay playback smoother (useful for autoplay, currently).
-        /// </summary>
-        public bool FrameAccuratePlayback;
+        private int currentFrameIndex;
 
-        public bool HasFrames => Frames.Count > 0;
+        private const double sixty_frame_time = 1000.0 / 60;
+
+        protected FramedReplayInputHandler(Replay replay)
+        {
+            // This replay frame ordering should be enforced on the Replay type
+            replay.Frames.Sort((x, y) => x.Time.CompareTo(y.Time));
```

Comment:
Is the intention to remove this eventually (in a follow up PR)? Probably best to prefix the comment with `TODO:`.

Also as this is done at construction time, it won't be applied to streaming frames so that's something to consider.